Providers Tutorial (#6095)
ethereum/web3.js: docs/docs/tutorials/providers.md

Before we get started, make sure you have a basic understanding of JavaScript and Ethereum. Additionally, we need to set up our environment by installing the following:

1. **Ganache**

    Ganache is a personal blockchain for Ethereum development that allows you to test how your smart contracts function in real-world scenarios. You can download it from [http://truffleframework.com/ganache](http://truffleframework.com/ganache).

2. **Node.js**

    Node.js is a JavaScript runtime environment that enables you to run JavaScript on the server-side. You can download it from [https://nodejs.org/en/download/](https://nodejs.org/en/download/).

3. **npm**

    npm (Node Package Manager) is used to publish and install packages to and from the public npm registry or a private npm registry. You can install it by following the instructions here: [https://docs.npmjs.com/downloading-and-installing-node-js-and-npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm).

    Alternatively, you can use **yarn** instead of **npm** by following the instructions here: [https://classic.yarnpkg.com/lang/en/docs/getting-started/](https://classic.yarnpkg.com/lang/en/docs/getting-started/).

> ## Types of Providers

Reviewer: ```suggestion
3. **npm**

    npm (Node Package Manager) is used to publish and install packages to and from the public npm registry or a private npm registry. You can install it by following the instructions here: [https://docs.npmjs.com/downloading-and-installing-node-js-and-npm](https://docs.npmjs.com/downloading-and-installing-node-js-and-npm).

    Alternatively, you can use **yarn** instead of **npm** by following the instructions here: [https://classic.yarnpkg.com/lang/en/docs/getting-started/](https://classic.yarnpkg.com/lang/en/docs/getting-started/).
    
4. **Geth** (Optional, used for IPC provider example)

    Geth (go-ethereum) is an Ethereum execution client meaning it handles transactions, deployment and execution of smart contracts and contains an embedded computer known as the Ethereum Virtual Machine. You can install it by following the instructions here: [https://geth.ethereum.org/docs/getting-started/installing-geth](https://geth.ethereum.org/docs/getting-started/installing-geth)

## Types of Providers
```

1. [HTTP Provider](#http-provider)
2. [WebSocket Provider](#websocket-provider)
3. [IPC Provider (for Node.js)](#websocket-provider)
4. [Third-party Providers (Compliant with EIP 1193)](#third-party-providers-compliant-with-eip-1193)

### HTTP Provider

The HTTP Provider allows you to connect to a publicly available Ethereum node, making it easy and straightforward to communicate with the Ethereum network from your web application.

To connect to the Ethereum network using the HTTP provider, follow these steps:

1. Open a command prompt or terminal window and navigate to the directory where you want to create the folder for this example.
2. Create a new folder and navigate to it:

    ```sh
    mkdir web3-providers-tutorial
    cd web3-providers-tutorial
    ```

3. Install Web3.js using npm:

    ```sh
    npm install web3@4.0.1-rc.1
    ```

4. Create a new JavaScript file called `web3-http-provider.js` in your code editor.

5. Copy and paste the following code into your `web3-http-provider.js` file and save it:

    ```js
    const { Web3 } = require('web3');

    // Connect to the Ethereum network using the HTTP provider

>     const ganacheUrl = 'http://localhost:7545';

Reviewer: ```suggestion
    const ganacheUrl = 'http://localhost:8545';
```

Author: Actually, the example uses Ganache application which runs on `7545` by default.

    		// Send a transaction to the network and wait for the transaction to be mined.
    		// Note that sending a transaction with Ganache will cause it, in its default configuration, to min a new block.
    		const transactionReceipt = await web3.eth.sendTransaction({
    			from: accounts[0],
    			to: accounts[1],
    			value: web3.utils.toWei('0.001', 'ether'),
    		});
    		console.log('Transaction Receipt:', transactionReceipt);

    		// Get the updated block number
    		const updatedBlockNumber = await web3.eth.getBlockNumber();
    		console.log('Updated block number:', updatedBlockNumber);
    	} catch (error) {
    		console.error('An error occurred:', error);
    	}
    }

    main();
    ```

6. Ensure that Ganache is running as mentioned in the [Prerequisites](#prerequisites) section.

7. In the command prompt or terminal window, type `node web3-http-provider.js` and press Enter. This will run your JavaScript file and connect to the Ethereum network using the HTTP provider and Ganache.

If everything is set up properly, you should see the current block number, a transaction hash, transaction receipt, and the updated block number printed in the console.

> This example showed how to connect to the Ethereum network using HTTP provider, get the current block number, send a transaction, and get the updated block number.

Reviewer: ```suggestion
If everything is set up properly, you should see the current block number, a transaction hash, transaction receipt, and the updated block number printed in the console:

```bash
Current block number: 0n
Transaction Receipt: {
  transactionHash: '0x0578672e97d072b4b91773c8bfc710e4f777616398b82b276323408e59d11362',
  transactionIndex: 0n,
  blockNumber: 1n,
  blockHash: '0x348a6706e7cce6547fae2c06b3e8eff1f58e4669aff88f0af7ca250ffdcdeef5',
  from: '0x6e599da0bff7a6598ac1224e4985430bf16458a4',
  to: '0x6f1df96865d09d21e8f3f9a7fba3b17a11c7c53c',
  cumulativeGasUsed: 21000n,
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  effectiveGasPrice: 2000000000n,
  type: 0n
}
Updated block number: 1n
```

This example showed how to connect to the Ethereum network using HTTP provider, get the current block number, send a transaction, and get the updated block number.
```

Reviewer: Formatting looks weird because Github's suggestions uses ` ```suggestion `, but I added the expected output:

```bash
Current block number: 0n
Transaction Receipt: {
  transactionHash: '0x0578672e97d072b4b91773c8bfc710e4f777616398b82b276323408e59d11362',
  transactionIndex: 0n,
  blockNumber: 1n,
  blockHash: '0x348a6706e7cce6547fae2c06b3e8eff1f58e4669aff88f0af7ca250ffdcdeef5',
  from: '0x6e599da0bff7a6598ac1224e4985430bf16458a4',
  to: '0x6f1df96865d09d21e8f3f9a7fba3b17a11c7c53c',
  cumulativeGasUsed: 21000n,
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  effectiveGasPrice: 2000000000n,
  type: 0n
}
Updated block number: 1n
```

### WebSocket Provider

WebSocket Provider allows us to communicate with the Ethereum node via WebSocket protocol, which is useful when we want continuous updates on our subscribed items. This provider is ideal for real-time applications that require constant updates from the Ethereum network.

Follow these steps to connect to the Ethereum network using WebSocket provider:

:::tip
The first 3 steps are the same as in the pervious section. So, you may skip them if you already executed the previous section.
:::

1. Open a command prompt or terminal window and navigate to where you would like to create the folder for this example.
2. Create a new folder and navigate to it:

    ```sh
    mkdir web3-providers-tutorial
    cd web3-providers-tutorial
    ```

3. Install Web3.js using npm:

    ```sh
     npm install web3@4.0.1-rc.1
    ```

4. Create a new JavaScript file called `web3-websocket-provider.js` in your code editor.

5. Copy and paste the following code into your `web3-websocket-provider.js` file and save it:

    ```js
    const { Web3 } = require('web3');

    // Connect to the Ethereum network using the WebSocket provider

>     const ganacheUrl = 'ws://localhost:7545';

Reviewer: ```suggestion
    const ganacheUrl = 'ws://localhost:8545';
```

Author: Actually, the example uses Ganache application which runs on `7545` by default.

    const wsProvider = new Web3.providers.WebsocketProvider(ganacheUrl);
    const web3 = new Web3(wsProvider);

    async function main() {
    	try {
    		console.log(

>     			'Do the provider supports subscription?:',

Reviewer: ```suggestion
    			'Does the provider support subscriptions?:',
```

    				}
    				if (error) {
    					console.log(error);
    				}
    				wsProvider.disconnect();
    			});
    		}, 2000);

    		// Get the list of accounts in the connected node which is in this case: Ganache.
    		const accounts = await web3.eth.getAccounts();

    		// Send a transaction to the network
    		const transactionReceipt = await web3.eth.sendTransaction({
    			from: accounts[0],
    			to: accounts[1],
    			value: web3.utils.toWei('0.001', 'ether'),
    		});
    		console.log('Transaction Receipt:', transactionReceipt);
    	} catch (error) {
    		console.error('An error occurred:', error);
    	}
    }

    main();
    ```

6. Ensure that Ganache is running as mentioned in the [Prerequisites](#prerequisites) section.

7. Type `node web3-websocket-provider.js` in the command prompt or terminal window and press Enter. This will run your JavaScript file.

If everything is set up properly, you should see the new block headers, transaction hash, and pending transaction printed in the console. The unique feature of WebSocket provider highlighted in this example is that it can subscribe to new block headers and pending transactions in real-time.

> ### IPC Provider (for Node.js)

Reviewer: ```suggestion
If everything is set up properly, you should see the new block headers, transaction hash, and pending transaction printed in the console. The unique feature of WebSocket provider highlighted in this example is that it can subscribe to new block headers and pending transactions in real-time:

```bash
Does the provider support subscriptions?: true
Transaction Receipt: {
  transactionHash: '0x0578672e97d072b4b91773c8bfc710e4f777616398b82b276323408e59d11362',
  transactionIndex: 0n,
  blockNumber: 1n,
  blockHash: '0x5c05248fe0fb8f45a8c9b9600904a36c0e5c74dce01495cfc72278c185fe7838',
  from: '0x6e599da0bff7a6598ac1224e4985430bf16458a4',
  to: '0x6f1df96865d09d21e8f3f9a7fba3b17a11c7c53c',
  cumulativeGasUsed: 21000n,
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  effectiveGasPrice: 2000000000n,
  type: 0n
}
```

### IPC Provider (for Node.js)
```

Reviewer: Added:

```bash
Does the provider support subscriptions?: true
Transaction Receipt: {
  transactionHash: '0x0578672e97d072b4b91773c8bfc710e4f777616398b82b276323408e59d11362',
  transactionIndex: 0n,
  blockNumber: 1n,
  blockHash: '0x5c05248fe0fb8f45a8c9b9600904a36c0e5c74dce01495cfc72278c185fe7838',
  from: '0x6e599da0bff7a6598ac1224e4985430bf16458a4',
  to: '0x6f1df96865d09d21e8f3f9a7fba3b17a11c7c53c',
  cumulativeGasUsed: 21000n,
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  effectiveGasPrice: 2000000000n,
  type: 0n
}
```

The IPC Provider allows you to connect to an Ethereum node using Inter-Process Communication (IPC) in a Node.js environment. This provider is useful when you have a local Ethereum node running on your machine and want to interact with it using Node.js.

> In the following steps you will run `geth` in development mode and you will run a code that reads geth accounts and send a transaction:

Reviewer: ```suggestion
In the following steps you will run `geth` in development mode and you will run code that reads Ethereum accounts and sends a transaction:
```

3. Create a new folder and navigate to it:

    ```sh
    mkdir web3-providers-tutorial
    cd web3-providers-tutorial
    ```

4. Install Web3.js using npm:

    ```sh
    npm install web3@4.0.1-rc.1
    ```

5. Create a new JavaScript file called `web3-ipc-provider.js` in your code editor.

6. Copy and paste the following code into your `web3-ipc-provider.js` file and save it:

    ```js
    const { Web3 } = require('web3');
    const { IpcProvider } = require('web3-providers-ipc');

    // Connect to the Ethereum network using IPC provider
    const ipcPath = '<path>'; // Replace with your actual IPC path
    const ipcProvider = new IpcProvider(ipcPath);

    const web3 = new Web3(ipcProvider);

    async function main() {
    	try {
    		console.log(
    			'Do the provider supports subscription?:',
    			ipcProvider.supportsSubscriptions(),
    		);

    		// Get the list of accounts in the connected node which is in this case: geth in dev mode.
    		const accounts = await web3.eth.getAccounts();
    		console.log('Accounts:', accounts);

    		// Send a transaction to the network
    		const transactionReceipt = await web3.eth.sendTransaction({
    			from: accounts[0],

>     			to: accounts[1],

Reviewer: ```suggestion
    			to: accounts[0],
```

Geth creates a single address by default. using `accounts[1]` for `to` results in `undefined` causing an error: `An error occurred: InvalidResponseError: Returned error: contract creation without any data provided`

    			value: web3.utils.toWei('0.001', 'ether'),
    		});
    		console.log('Transaction Receipt:', transactionReceipt);
    	} catch (error) {
    		console.error('An error occurred:', error);
    	}
    }

    main();
    ```

7. replace `<path>` with the `ipcPath` that you had specified, when starting the `geth` node, in the first step.

    **Note**: replace `<path>` with the `ipcPath` specified when starting the `geth` node.

> 8. Type `node web3-ipc-provider.js` in the command prompt or terminal window and press Enter. This will run your JavaScript file.

Reviewer: ```suggestion
7. replace `<path>` with the `ipcPath` that you had specified, when starting the `geth` node, in the first step.

8. Type `node web3-ipc-provider.js` in the command prompt or terminal window and press Enter. This will run your JavaScript file.
```

Reviewer: I removed

```
**Note**: replace `<path>` with the `ipcPath` specified when starting the `geth` node.
```

because it's repeating the information given right above it

Author: Thanks, I forgot to remove it.

If everything is set up properly, you should see the list of accounts and transaction receipt printed in the console.

> Keep in mind that using IPC Provider with `geth` in development mode in a production environment is not recommended as it can pose a security risk.

Reviewer: ```suggestion
If everything is set up properly, you should see the list of accounts and transaction receipt printed in the console:

```bash
Do the provider supports subscription?: true
Accounts: [ '0x82333ED0FAA7a883297C4d8e0FDE1E1CFABAeB7D' ]
Transaction Receipt: {
  blockHash: '0xd1220a9b6f86083e420da025179593f5aad3732165a687019a89528a4ab2bcd8',
  blockNumber: 1n,
  cumulativeGasUsed: 21000n,
  effectiveGasPrice: 1000000001n,
  from: '0x82333ed0faa7a883297c4d8e0fde1e1cfabaeb7d',
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  to: '0x82333ed0faa7a883297c4d8e0fde1e1cfabaeb7d',
  transactionHash: '0x76c05df78dc5dbfade0d11322b3cadc894c17efe36851856aca29488b47c3fbd',
  transactionIndex: 0n,
  type: 0n
}
```

Keep in mind that using IPC Provider with `geth` in development mode in a production environment is not recommended as it can pose a security risk.
```

Reviewer: Added:

```bash
Do the provider supports subscription?: true
Accounts: [ '0x82333ED0FAA7a883297C4d8e0FDE1E1CFABAeB7D' ]
Transaction Receipt: {
  blockHash: '0xd1220a9b6f86083e420da025179593f5aad3732165a687019a89528a4ab2bcd8',
  blockNumber: 1n,
  cumulativeGasUsed: 21000n,
  effectiveGasPrice: 1000000001n,
  from: '0x82333ed0faa7a883297c4d8e0fde1e1cfabaeb7d',
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  to: '0x82333ed0faa7a883297c4d8e0fde1e1cfabaeb7d',
  transactionHash: '0x76c05df78dc5dbfade0d11322b3cadc894c17efe36851856aca29488b47c3fbd',
  transactionIndex: 0n,
  type: 0n
}
```

Here is a step-by-step example and a code snippet to connect your web application to the Ethereum network using `@truffle/hdwallet-provider` as an example of an external provider compliant with EIP 1193.

1. Open a command prompt or terminal window in a new folder.
2. Type `npm init -y` and press Enter. This will create a `package.json` file in the current directory.
3. Install Web3.js and HTTP provider using npm:

    ```sh
    npm install web3@4.0.1-rc.1 @truffle/hdwallet-provider bip39
    ```

4. Create a new JavaScript file, called `index.js`, in your code editor.
5. Copy and paste the following code into your JavaScript file, and then save the file:

    ```js
    const { Web3 } = require('web3');
    const HDWalletProvider = require('@truffle/hdwallet-provider');
    const bip39 = require('bip39');

>     const mnemonic = bip39.generateMnemonic(); //generates seed phrase

Reviewer: ```suggestion
    const mnemonic = bip39.generateMnemonic(); // generates seed phrase
```

    console.log('seed phrase:', mnemonic);

    // Connect to the Ethereum network using an HTTP provider and WalletProvider
    const provider = new HDWalletProvider(
    	mnemonic,
    	'https://mainnet.infura.io/v3/YOUR_INFURA_PROJECT_ID',
    );
    const web3 = new Web3(provider);

    // Get the current block number from the network
    web3.eth
    	.getBlockNumber()
    	.then(function (blockNumber) {
    		console.log('Current block number:', blockNumber);
    	})
    	.catch(function (error) {
    		console.log('Error:', error);
    	});
    ```

> 6. Replace `'YOUR_INFURA_PROJECT_ID'` with your own Infura project ID. You can obtain an Infura project ID by signing up for a free account at https://infura.io/register. Alternatively, you can use any other URL that is compatible with HDWalletProvider, such as a local Ganache accessible at `'http://localhost:7545'`.

Reviewer: ```suggestion
6. Replace `'YOUR_INFURA_PROJECT_ID'` with your own Infura project ID. You can obtain an Infura project ID by signing up for a free account at https://infura.io/register. Alternatively, you can use any other URL that is compatible with HDWalletProvider, such as a local Ganache accessible at `'http://localhost:8545'`.
```

Author: Actually, the example uses Ganache application which runs on `7545` by default.

7. In the command prompt, run `node index.js` and press Enter. This will execute your JavaScript file and connect to the Ethereum network using HDWalletProvider with Infura.

If everything is set up properly, you should see the current block number printed in the console.

> This code will connect your web application to the Ethereum network using truffle HD Wallet-enabled Web3 provider. You can modify it to interact with the network, perform transactions, and read/write data from the Ethereum network.

Reviewer: ```suggestion
7. In the command prompt, run `node index.js` and press Enter. This will execute your JavaScript file and connect to the Ethereum network using HDWalletProvider with Infura.

If everything is set up properly, you should see the current block number printed in the console:

:::danger
Your seed phrase gives complete access to your Ethereum account and it should **never** be shared with anyone you don't want to give full access to your account. The seed phrase is `console.log`ed in the code example to show you what it looks like, but you should **never** do this with a seed phrase to an account you plan on using to send real money
:::

```bash
seed phrase: remain climb clock valid budget cable tunnel force split level measure repair
Current block number: 17317844n
```

This code will connect your web application to the Ethereum network using truffle HD Wallet-enabled Web3 provider. You can modify it to interact with the network, perform transactions, and read/write data from the Ethereum network.
```

Reviewer: Added:

```bash
:::danger
Your seed phrase gives complete access to your Ethereum account and it should **never** be shared with anyone you don't want to give full access to your account. The seed phrase is `console.log`ed in the code example to show you what it looks like, but you should **never** do this with a seed phrase to an account you plan on using to send real money
:::

seed phrase: remain climb clock valid budget cable tunnel force split level measure repair
Current block number: 17317844n
```

Follow these steps to connect to the Ethereum network with MetaMask and Web3.js, including the steps to create a local web server using Node.js:

1. Ensure that Ganache is running as mentioned in the [Prerequisites](#prerequisites) section.
2. Install the MetaMask extension for your browser. You can download MetaMask from their website: https://metamask.io/.
3. Create a new HTML file in your code editor.
4. Copy and paste the following code into your HTML file:

    ```html
    <!DOCTYPE html>
    <html lang="en">
    	<head>
    		<meta charset="UTF-8" />
    		<title>Connecting to the Ethereum network with Web3.js and MetaMask</title>
    	</head>
    	<body>
    		<h1>Connecting to the Ethereum network with Web3.js and MetaMask</h1>
    		<div id="log">
    			You need to approve connecting this website to MetaMask. Click on the MetaMask icon
    			in the browser extension, if it did not show a popup already.
    		</div>

>     		<script src="https://cdn.jsdelivr.net/npm/web3@4.0.1-rc.1/dist/web3.min.js"></script>

Reviewer: This doesn't currently exist

Reviewer: https://github.com/web3/web3.js/issues/6117

    						'Please install MetaMask to connect to the Ethereum network.';
    				}
    			});
    		</script>
    	</body>
    </html>
    ```

    5. Save the file as `index.html`.
    6. Open a terminal or command prompt window.
    7. Navigate to the directory where your HTML file is located.
    8. Type `npm init -y` and press Enter. This will create a `package.json` file in the current directory.
    9. Type `npm install --save express` and press Enter. This will install the Express module and add it to your project's dependencies.
    10. Create a new file called `server.js` in the same directory as your HTML file.
    11. Copy and paste the following code into `server.js`:

    ```js
    const express = require('express');
    const app = express();
    const path = require('path');

    app.use(express.static(path.join(__dirname, '.')));

    app.listen(8000, () => {
    	console.log('Server started on port 8000');
    });
    ```

5. Save the file.
6. Open your command prompt window and navigate to the directory where `server.js` is located.
7. Type `node server.js` and press Enter. This will start the server on port 8000.

> 8. Open your web browser and navigate to `http://localhost:8000/`. MetaMask should ask for your approval to connect to your website. Follow the steps and give your consent.

Reviewer: We could also just have the user open the HTML file in their browser and forego having to install and setup express

Author: Unfortunately, this would not work because of browser security restrictions. 
(I discovered that I had a very old question about this https://github.com/MetaMask/metamask-extension/issues/6162 :smile:)

This will start a `geth` node in development mode with IPC enabled and an IPC path specified. If the command is successful, the `geth` node will be running, and you should see output similar to the following:

```sh
INFO [12-10|15:10:37.121] IPC endpoint opened              url=<path>
INFO [12-10|15:10:37.122] HTTP endpoint opened             url=http://localhost:8545
INFO [12-10|15:10:37.122] WebSocket endpoint opened        url=ws://localhost:8546
INFO [12-10|15:10:37.127] Mapped network port              proto=udp extport=0 intport=30303 interface=UPnP(UDP)
```

2. Open a command prompt or terminal window and navigate to where you would like to create the folder for this example.
3. Create a new folder and navigate to it:

    ```sh
    mkdir web3-providers-tutorial
    cd web3-providers-tutorial
    ```

4. Install Web3.js using npm:

    ```sh
    npm install web3
    ```

5. Create a new JavaScript file called `web3-ipc-provider.js` in your code editor.
6. Copy and paste the following code into your `web3-ipc-provider.js` file and save it:

    ```js
    const { Web3 } = require('web3');

    // Connect to the Ethereum network using IPC provider
    const ipcPath = '<path>';

>     const ipcProvider = new Web3.providers.IpcProvider(ipcPath);

Reviewer 2: this code will need update as IPC is not exported through main web3 package now.

```bash
Current block number: 0n
Transaction Receipt: {
  transactionHash: '0x0578672e97d072b4b91773c8bfc710e4f777616398b82b276323408e59d11362',
  transactionIndex: 0n,
  blockNumber: 1n,
  blockHash: '0x348a6706e7cce6547fae2c06b3e8eff1f58e4669aff88f0af7ca250ffdcdeef5',
  from: '0x6e599da0bff7a6598ac1224e4985430bf16458a4',
  to: '0x6f1df96865d09d21e8f3f9a7fba3b17a11c7c53c',
  cumulativeGasUsed: 21000n,
  gasUsed: 21000n,
  logs: [],
  logsBloom: '0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000',
  status: 1n,
  effectiveGasPrice: 2000000000n,
  type: 0n
}
Updated block number: 1n

> ```
> 
> ### WebSocket Provider

Reviewer 3: ```suggestion
```

This example showed how to connect to the Ethereum network using HTTP provider, get the current block number, send a transaction, and get the updated block number.

### WebSocket Provider
```

> 
> ### WebSocket Provider

Reviewer: ```suggestion

This example showed how to connect to the Ethereum network using HTTP provider, get the current block number, send a transaction, and get the updated block number.

### WebSocket Provider
```

Reviewer: You had this block of text before, but I think it got mistakenly removed with the edits

Author: Actually, I removed it as it is a repetition of what is mentioned here: 
> To connect to the Ethereum network using the HTTP provider, follow these steps:

And here:
> If everything is set up properly, you should see the current block number, the transaction receipt, and the updated block number printed in the console:

Thanks,
